File: trunk/src/cgi/msg_cgi/cgi_msg_get_session_list.cpp
#include "cgi_msg_server.h"
#include "proto_io_tcp_client.h"
#include <time.h>
#include "msg.pb.h"
#include "user_util.h"

class CgiMsgGetSessionList: public CgiMsgServer
{
public:
	CgiMsgGetSessionList():
		CgiMsgServer(0, "config.ini", "logger.properties")
	{

	}

	bool InnerProcess()
	{
		//1.get HTTP params
		uint64_t openid_md5 = strtoul(((string)GetInput().GetValue("openid_md5")).c_str(), NULL, 10);

		if(0 == openid_md5)
		{
			DoReply(CGI_RET_CODE_INVALID_PARAM);
			LOG4CPLUS_ERROR(logger, "INVALID OPENID MD5");
			return true;	
		}		

		//2.Build requestMsg protobuf
		std::string strErrMsg = "";
		::hoosho::msg::Msg stRequestMsg;
		::hoosho::msg::Msg stResponseMsg;
		
		::hoosho::msg::MsgHead* stRequestMsgHead = stRequestMsg.mutable_head();
		stRequestMsgHead->set_cmd(::hoosho::msg::QUERY_MSG_SESSION_LIST_REQ);		
		stRequestMsgHead->set_seq(time(NULL));

		::hoosho::sixin::QuerySixinSessionListReq* stMsgQueryReq = stRequestMsg.mutable_msg_query_session_list_req();
		stMsgQueryReq->set_openid_md5(openid_md5);		

		//3.send to server, and recv responseMsg protobuf
		common::protoio::ProtoIOTcpClient ioclient(msgServerIP, msgServerPort);

		int iRet;
		iRet = ioclient.io(stRequestMsg, stResponseMsg, strErrMsg);

		if(iRet != 0)
		{
			LOG4CPLUS_ERROR(logger, "ProtoIOTcpClient IO failed, errmsg = " << strErrMsg);
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			return true;
		}

		LOG4CPLUS_DEBUG(logger, "response Msg: \n" << stResponseMsg.Utf8DebugString());		
		//parse response
		const ::hoosho::msg::MsgHead& stResponseMsgHead = stResponseMsg.head();
		if(stResponseMsgHead.cmd() != ::hoosho::msg::QUERY_MSG_SESSION_LIST_RES)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.cmd="<<stResponseMsgHead.cmd()<<", unknown, fuck!!!");
			return true;	
		}

		if(stResponseMsgHead.result() != ::hoosho::msg::E_OK)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.result="<<stResponseMsgHead.result());
			return true;	
		}

		//4.Build strResponse Json from responseMsg protobuf
		const ::hoosho::sixin::QuerySixinSessionListRes& stMsgQueryRes = stResponseMsg.msg_query_session_list_res();


		//5.Get Userinfo by openid from user server		
		::hoosho::msg::Msg stRequest;
		::hoosho::msg::MsgHead* pHead = stRequest.mutable_head();
		pHead->set_cmd(::hoosho::msg::QUERY_USER_DETAIL_INFO_REQ);
		pHead->set_seq(time(NULL));
		::hoosho::user::QueryUserDetailInfoReq* pQueryUserDetailInfoReq = stRequest.mutable_query_user_detail_info_req();
		pQueryUserDetailInfoReq->add_openid_md5_list(openid_md5);
		for(int i=0; i<stMsgQueryRes.session_size(); i++)
		{
			const ::hoosho::commstruct::Session& session = stMsgQueryRes.session(i);	
			if(openid_md5 == session.openid_md5_from())
			{
				pQueryUserDetailInfoReq->add_openid_md5_list(session.openid_md5_to());
			}
			else
			{
				pQueryUserDetailInfoReq->add_openid_md5_list(session.openid_md5_from());
			}
		}
		

		//io
		strErrMsg = "";
		::hoosho::msg::Msg stResponse;
		::common::protoio::ProtoIOTcpClient stProtoIOTcpClient(userServerIP, userServerPort);
		iRet = stProtoIOTcpClient.io(stRequest, stResponse, strErrMsg);
		if(iRet < 0)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "Process failed, stProtoIOTcpClient.io, strErrMsg="<<strErrMsg);
			return true;	
		}

		LOG4CPLUS_DEBUG(logger, "response Msg: \n" << stResponse.Utf8DebugString());		
		//parse response
		const ::hoosho::msg::MsgHead& stHead = stResponse.head();
		if(stHead.cmd() != ::hoosho::msg::QUERY_USER_DETAIL_INFO_RES)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.cmd="<<stHead.cmd()<<", unknown, fuck!!!");
			return true;	
		}

		if(stHead.result() != ::hoosho::msg::E_OK)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.result="<<stHead.result());
			return true;	
		}

		lce::cgi::CAnyValue stHostUserInfo;
		const ::hoosho::commstruct::UserDetailInfo& stHostUserDetailInfo = stResponse.query_user_detail_info_res().user_detail_info_list(0);
		UserInfoPB2Any(stHostUserDetailInfo, stHostUserInfo);
		
		lce::cgi::CAnyValue session_list;		
		for(int i=stMsgQueryRes.session_size()-1; i>=0; i--)
		{			
			const ::hoosho::commstruct::Session& session = stMsgQueryRes.session(i);	
			bool flag = false;
			
			lce::cgi::CAnyValue stGuestUserInfo;
			for(int j=stResponse.query_user_detail_info_res().user_detail_info_list_size()-1; j>=0; j--)
			{
				const ::hoosho::commstruct::UserDetailInfo& stGuestUserDetailInfo = stResponse.query_user_detail_info_res().user_detail_info_list(j);
				if(openid_md5 == session.openid_md5_from())
				{
					if(stGuestUserDetailInfo.openid_md5() != session.openid_md5_to())
						continue;
					else
					{
						UserInfoPB2Any(stGuestUserDetailInfo, stGuestUserInfo);	
						flag = true;
						break;
					}	
				}
				else if(openid_md5 == session.openid_md5_to())
				{
					if(stGuestUserDetailInfo.openid_md5() != session.openid_md5_from())
						continue;
					else
					{
						UserInfoPB2Any(stGuestUserDetailInfo, stGuestUserInfo);	
						flag = true;
						break;
					}	
				}								
			}

			if(!flag)
				continue;			

			lce::cgi::CAnyValue item;

			item["openid_md5_from"] = int_2_str(session.openid_md5_from());
			item["openid_md5_to"] = int_2_str(session.openid_md5_to());
			
			if(openid_md5 == session.openid_md5_from())
			{
				item["userinfo_from"] = stHostUserInfo;
				item["userinfo_to"] = stGuestUserInfo;
			}
			else
			{
				item["userinfo_from"] = stGuestUserInfo;
				item["userinfo_to"] = stHostUserInfo;
			}
						
			item["content"] = session.content();
			item["create_ts"] = int_2_str(session.create_ts());
			item["newmsg_status"] = session.newmsg_status();

			session_list.push_back(item);			
		}
		GetAnyValue()["session_list"] = session_list;		
		
		DoReply(CGI_RET_CODE_OK);
		return true;		
	}
		
};

int main()
{
	CgiMsgGetSessionList cgi;
	if(!cgi.Run())
	{
		return -1;
	}
	return 0;
}

File: trunk/src/cgi_python/msg_cgi/cgi_msg_get_session_list.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

import sys
import time

# 包含上级目录，即../common
sys.path.append('..')

# cgi, proto_io
from common.cgi.cgi_ret_code_def import CGI_RET_CODE
import common.cgi.cgi_my as _cgi
import common.proto_io.proto_io_tcp_client as _tcp_client

# proto
import common.proto.msg_pb2 as _msg
import common.proto.msg_sixin_pb2 as _msg_sixin
import common.proto.comm_struct_pb2 as _comm_struct
import common.proto.comm_enum_pb2 as _comm_enum

# util
from common.util.common_util import *

# msgServerIP = '127.0.0.1'

def FetchUserInfo(_listOpenidMd5, _pbUserInfoList):
	# 1.Build requestMsg protobuf
	stUserInfoRequestMsg = _msg.Msg()

	stUserInfoRequestMsgHead = stUserInfoRequestMsg.head
	stUserInfoRequestMsgHead.cmd = _msg.QUERY_USER_DETAIL_INFO_REQ
	stUserInfoRequestMsgHead.seq = int(time.time())

	stQueryUserDetailInfoReq = stUserInfoRequestMsg.query_user_detail_info_req
	for openidMd5 in _listOpenidMd5:
		stQueryUserDetailInfoReq.openid_md5_list.append(openidMd5)

	# 2.Send to server, and recv responseMsg protobuf
	ioclient = _tcp_client.ProtoIOTcpClient(userServerIP, userServerPort)
	dictMsg = {'requestMsg':stUserInfoRequestMsg, 'responseMsg':None, 'strErrMsg':None}

	ret = ioclient.io(dictMsg)

	if ret != 0:	
		logger.error('FetchFeedsInfo IO failed, errmsg = %s', dictMsg['strErrMsg'])
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return -1

	stUserInfoResponseMsg = dictMsg['responseMsg']
	logger.debug('responseMsg: \n %s', stUserInfoResponseMsg)

	if stUserInfoResponseMsg.head.cmd != _msg.QUERY_USER_DETAIL_INFO_RES:
		logger.error('FetchUserInfo response.head.cmd=%d, unkown, fuck!!!', stUserInfoResponseMsg.head.cmd)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return -1

	if stUserInfoResponseMsg.head.result != _msg.E_OK:
		logger.error('FetchUserInfo response.head.result=%d, not E_OK!', stUserInfoResponseMsg.head.result)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return -1

	# 3.Parse	
	res = stUserInfoResponseMsg.query_user_detail_info_res

	for pbUserInfo in res.user_detail_info_list:
		_pbUserInfoList.append(pbUserInfo)
	
	return 0

def InnerProcess():
	# 1.Get HTTP params
	openid_md5 = int(mycgi.getEnvValue('openid_md5', 0))

	# 2.Params check
	if 0 == openid_md5:
		logger.error('INVALID PARAM!')
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_INVALID_PARAM)		
		return True

	# 3.Build requestMsg protobuf
	stRequestMsg = _msg.Msg()

	stRequestMsgHeader = stRequestMsg.head
	stRequestMsgHeader.cmd = _msg.QUERY_MSG_SESSION_LIST_REQ
	stRequestMsgHeader.seq = int(time.time())

	stRequestMsgBody = stRequestMsg.msg_query_session_list_req
	stRequestMsgBody.openid_md5 = openid_md5

	logger.debug('requestMsg: \n %s', stRequestMsg)

	# 4.Send to server, and recv responseMsg protobuf	
	ioclient = _tcp_client.ProtoIOTcpClient(msgServerIP, msgServerPort)

	dictMsg = {'requestMsg':stRequestMsg, 'responseMsg':None, 'strErrMsg':None}

	ret = ioclient.io(dictMsg)

	if ret != 0:	
		logger.error('ProtoIOTcpClient IO failed, errmsg = %s', dictMsg['strErrMsg'])
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	stResponseMsg = dictMsg['responseMsg']
	logger.debug('responseMsg: \n %s', stResponseMsg)

	if stResponseMsg.head.cmd != _msg.QUERY_MSG_SESSION_LIST_RES:
		logger.error('response.head.cmd=%d, unkown, fuck!!!', stResponseMsg.head.cmd)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	if stResponseMsg.head.result != _msg.E_OK:
		logger.error('response.head.result=%d, not E_OK!', stResponseMsg.head.result)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	# 5.Get UserInfo
	listOpenidMd5 = []
	listOpenidMd5.append(openid_md5)
	stMsgQueryRes = stResponseMsg.msg_query_session_list_res
	for session in stMsgQueryRes.session:
		if openid_md5 == session.openid_md5_from:
			listOpenidMd5.append(session.openid_md5_to)
		else:
			listOpenidMd5.append(session.openid_md5_from)

	pbUserInfoList = []
	if FetchUserInfo(listOpenidMd5, pbUserInfoList) < 0:
		logger.error('FetchUserInfo failed!!!')
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)
		return True

	stHostUserInfo = UserInfoPB2Dict(pbUserInfoList[0])

	session_list = []
	for i in range(len(stMsgQueryRes.session))[::-1]:
		tSession = stMsgQueryRes.session[i]
		flag = False

		for j in range(len(pbUserInfoList))[::-1]:
			stGuestUserInfo = None
			tUserInfo = pbUserInfoList[j]
			if openid_md5 == tSession.openid_md5_from:
				if tUserInfo.openid_md5 == tSession.openid_md5_to:
					continue
				else:
					stGuestUserInfo = UserInfoPB2Dict(tUserInfo)
					flag = True
					break
			elif openid_md5 == tSession.openid_md5_to:
				if tUserInfo.openid_md5 == tSession.openid_md5_from:
					continue
				else:
					stGuestUserInfo = UserInfoPB2Dict(tUserInfo)
					flag = True
					break

		if flag is not True:
			continue

		dictItem = {}
		dictItem['openid_md5_from'] = str(tSession.openid_md5_from)
		dictItem['openid_md5_to'] = str(tSession.openid_md5_to)

		if openid_md5 == tSession.openid_md5_from:
			dictItem['userinfo_from'] = stHostUserInfo
			dictItem['userinfo_to'] = stGuestUserInfo
		else:
			dictItem['userinfo_from'] = stGuestUserInfo
			dictItem['userinfo_to'] = stHostUserInfo

		dictItem['content'] = tSession.content
		dictItem['create_ts'] = str(tSession.create_ts)
		dictItem['newmsg_status'] = tSession.newmsg_status

		session_list.append(dictItem)

	mycgi.addBody('session_list', session_list)

	mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_OK)
	return True



if __name__ == '__main__':
	# __file__ 获取模块所在路径
	mycgi = _cgi.myCgi(__file__, 'logging.conf')

	# logger	
	logger = mycgi.getLogger()

	InnerProcess()	